src/Language/PureScript/Linter/Exhaustive.hs

```diff
@@ -0,0 +1,228 @@
+--
+completeMissing :: Ord a =>
+  (a -> Maybe b -> Maybe c -> d) -> {- b or c might be missing -}
+  [(a, b)] -> {- Collection of bs -}
+  [(a, c)] -> {- Collection of cs -}
+  [d] {- Zipped result -}
+completeMissing _ [] [] = []
+completeMissing f bs [] = map (\(s, b) -> f s (Just b) Nothing) bs
+completeMissing f [] bs = map (\(s, b) -> f s Nothing (Just b)) bs
+completeMissing f bsl@((s, b):bs) bsr@((s', b'):bs')
+  | s < s' = (f s (Just b) Nothing) : completeMissing f bs bsr
+  | s > s' = (f s' Nothing (Just b')) : completeMissing f bsl bs'
+  | otherwise = (f s (Just b) (Just b')) : completeMissing f bs bs'
+
+-- |
+-- Find the uncovered set between two binders:
+-- the first binder is the case we are trying to cover
+-- the second one is the matching binder
+--
+missingCasesSingle :: Environment -> Binder -> Binder -> [Binder]
+missingCasesSingle _ _ NullBinder = []
+missingCasesSingle _ _ (VarBinder _) = []
+missingCasesSingle env (VarBinder _) b = missingCasesSingle env NullBinder b
+missingCasesSingle env NullBinder cb@(ConstructorBinder con _) =
+  concatMap (\cp -> missingCasesSingle env cp cb) allPatterns
+  where
+  allPatterns = map (\(p, t) -> ConstructorBinder (qualifyProperName p con) (initialize $ length t)) $ getConstructors env con
+missingCasesSingle env cb@(ConstructorBinder con bs) (ConstructorBinder con' bs')
+  | con == con' = map (ConstructorBinder con) (missingCasesMultiple env bs bs')
+  | otherwise = [cb]
+missingCasesSingle _ NullBinder (ArrayBinder bs)
+  | null bs = [] 
+  | otherwise = []
+missingCasesSingle env NullBinder (ObjectBinder bs) =
+  map (ObjectBinder . zip (map fst bs)) allMisses
+  where
+  allMisses = missingCasesMultiple env (initialize $ length bs) (map snd bs)
+missingCasesSingle env (ObjectBinder bs) (ObjectBinder bs') =
+  map (ObjectBinder . zip sortedNames) $ uncurry (missingCasesMultiple env) (unzip binders)
+  where
+  sortNames = sortBy (compare `on` fst)
+
+  (sbs, sbs') = (sortNames bs, sortNames bs')
+
+  compB :: a -> Maybe a -> Maybe a -> (a, a)
+  compB e b b' = (fm b, fm b')
+    where
+    fm = fromMaybe e
+
+  compBS :: Eq a => b -> a -> Maybe b -> Maybe b -> (a, (b, b))
+  compBS e s b b' = (s, compB e b b')
+
+  (sortedNames, binders) = unzip $ completeMissing (compBS NullBinder) sbs sbs'
+missingCasesSingle _ NullBinder (BooleanBinder b) = [BooleanBinder $ not b]
+missingCasesSingle _ (BooleanBinder bl) (BooleanBinder br) = [BooleanBinder $ bl == br]
+missingCasesSingle env b (PositionedBinder _ _ cb) = missingCasesSingle env b cb
+missingCasesSingle _ b _ = [b]
+
+-- |
+-- Returns the uncovered set of binders
+-- the first argument is the list of uncovered binders
+-- the second argument is a clause of a pattern matching def
+--
+missingCasesMultiple :: Environment -> [Binder] -> [Binder] -> [[Binder]]
+missingCasesMultiple env = go
+  where
+  go [] _ = []
+  go (x:xs) (y:ys)
+    | null miss = map (x :) (go xs ys)
+    | otherwise = map (: xs) miss ++ map (y :) (go xs ys)
```

Comment:
Let's document this line, since I was intially confused why `(y :)` was adequate :)